### PR TITLE
Issue 428 kohei

### LINKED ISSUE #428: Padding is not zero
Padding is playing an important role in `tokens_ngrams()`. If paddings are zero, there will be not ngrams with of empty strings.
```r
toks <- tokens('a b c d e f g')
toks2 <- tokens_remove(toks, c('b', 'e'), padding = FALSE)
toks3 <- tokens_remove(toks, c('b', 'e'), padding = TRUE)
tokens_ngrams(toks)
# tokens from 1 document.
# Component 1 :
# [1] "a_b" "b_c" "c_d" "d_e" "e_f" "f_g"

tokens_ngrams(toks2)
# tokens from 1 document.
# Component 1 :
# [1] "a_c" "c_d" "d_f" "f_g"

tokens_ngrams(toks3)
# tokens from 1 document.
# Component 1 :
# [1] "a_"  "_c"  "c_d" "d_"  "_f"  "f_g"

unclass(toks3)
# [[1]]
# [1] 2 1 3 4 1 5 6
# 
# attr(,"types")
# [1] ""  "a" "c" "d" "f" "g"
# attr(,"what")
# [1] "word"
# attr(,"ngrams")
# [1] 1
# attr(,"concatenator")
# [1] ""
```
There was a set of code to keep padding = 0 in dev_lookup branch.

### PR DESCRIPTION
**Pull request by koheiw**
This should fix #428 without side effects

### CODE REVIEW

**Comment by kbenoit** [comment]
Might be a separate issue from #428 but I noticed this when adding the tests to `test-tokens.R`:

```r
    toks <- tokens(c(doc1 = 'a b c d e f g'))
    toks2 <- tokens_remove(toks, c('b', 'e'), padding = FALSE)
    expect_equal(as.list(tokens_ngrams(toks2, n = 2, skip = 2)),
                 list(doc1 = c("a_d", "b_e", "c_f", "d_g")))
```
fails, since:
```r
tokens_ngrams(toks2, n = 2, skip = 2)
# tokens from 1 document.
# doc1 :
# [1] "a_f" "c_g"
```
Bug?

**Comment by koheiw** [comment]
```
> toks2
tokens from 1 document.
doc1 :
[1] "a" "c" "d" "f" "g"
```
a_f and c_g are correct.

**Comment by koheiw** [comment]
By the way, ngram is working with padding
```
> toks3 <- tokens_remove(toks, c('b', 'e'), padding = TRUE)
> toks3
tokens from 1 document.
doc1 :
[1] "a" ""  "c" "d" ""  "f" "g"
> tokens_ngrams(toks3, n = 2, skip = 2)
tokens from 1 document.
doc1 :
[1] "a_d" "c_f" "d_g"

```

**Comment by kbenoit** [comment]
Now need to fix for dfm:
```r
test_that("test dfm with padded tokens, padding = TRUE", {
    toks <- tokens(c(doc1 = 'a b c d e f g',
                     doc2 = 'a b c g'))
    toks3 <- tokens_remove(toks, c('b', 'e'), padding = TRUE)
    expect_equivalent(as.matrix(dfm(toks3)),
                      matrix(c(1, 1, 1, 1, 1, 0, 1, 0, 1, 1), nrow = 2))
    ## and there should be a feature called "" with counts = 2, 1
})
```
I suggest that there should be a token called "" added to the dfm if there is padding in the tokens object. So we would need to shift the index right and add "" to the front.

**Comment by koheiw** [comment]
Agree. Should I do this?

**Comment by kbenoit** [comment]
yes please

**Comment by kbenoit** [comment]
We also test token_lookup/wordstem/tolower/toupper/select/remove, as.character.tokens, fcm, sequences, ntype, ntoken, nsyllable on tokens that include padding.

File: R/nfunctions.R [hunk 1]
```diff
@@ -73,6 +73,17 @@ nfeature.dfm <- function(x) {
     ncol(x)
 }
 
+#' @noRd
+#' @export
+nfeature.tokens <- function(x) {
+    if (0 %in% unlist(x, use.names = FALSE)) {
+        length(types(x)) + 1
+    }else{
```
Comment (kbenoit) [inline R/nfunctions.R]: Spacing!!

**Comment by kbenoit** [comment]
I just added tests for `tokens_remove` and `tokens_select` on padded `x` when the features to remove or select include `""`. Remove works, select not yet:

```r
> test_that("tokens_select works on \"\" with tokens containing padding", {
+     toks <- tokens(c(doc1 = 'a b c d e f g'))
+     toks <- tokens_remove(toks, c('b', 'e'), padding = TRUE)
+     expect_equal(as.character(tokens_select(toks, c("a", "b", ""))),
+                  c("a", "", ""))
+ })
Error: Test failed: 'tokens_select works on "" with tokens containing padding'
* as.character(tokens_select(toks, c("a", "b", ""))) not equal to c("a", "", "").
Lengths differ: 1 vs 3
> tokens_select(toks, c("a", "b", ""))
tokens from 1 document.
doc1 :
[1] "a"

``

**Pull request merged**